apiclient/rancher_api/api.py

```diff
@@ -56,8 +56,14 @@ def cluster_version(self):
         if not self._version:
             code, data = self.settings.get('server-version')
             ver = data['value']
-            # XXX: fix master-xxx-head to 8.8.8, need the API fix the problem
-            self._version = parse_version('8.8.8' if 'master' in ver else ver)
+
+            # Rancher versions look like v2.9.2, or v2.9.2-something
```

Comment:
Although Rancher is not using `master-head`, we would still need to consider cases like:
1. `v2.9.1-rc6`
2. `v2.9-2f9a7509687f559d772be75d61df7f7409803883-head`

As the second case would be the latest of `v2.9`, we need to fill the patch version for it. (to notice that there will not have something like `v2.9.1-sha-head`)

so, I would suggest that we could update it to something like
https://github.com/harvester/tests/blob/4d4d22b397e4fa8ff73ce193c6fce6b78624d356/apiclient/harvester_api/api.py#L43-L55

which we already implemented in `HarvesterAPI`.
To notice that Python usually use try-except to control the flow called EAFP.[^1]

[^1]: https://devblogs.microsoft.com/python/idiomatic-python-eafp-versus-lbyl/

Reply:
That was what I did initially, but @albinsun wasn't happy with it.

Reply:
I was talking about the order but not drop error handling.
My thought was
1. Existing implementation works for most cases
   (Except head form like `v2.9-2f9a7509687f559d-head`)
2. I thought you are dealing the head form case, so it's natually to suggest
    ```
    try:
       # Existing implementation
    except:
       # Dealing exceptions
    ```

Anyway, per discussed, It's reasonable for me to align with Harvester API side implementation.

Reply:
I think I have a good implementation now. It tries hard to parse the version, it keeps the suffix information for later reference and if all fails, it behaves like the Harvester API implementation.
I noticed that with my version of the `pkg_resources` library `parse_version` never raises an exception, but will produce a `LegacyVersion` object in case it doesn't know how to deal with it:
```
Python 3.6.15 (default, Sep 23 2021, 15:41:43) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pkg_resources import parse_version
>>> parse_version("foobar-noversion")
<LegacyVersion('foobar-noversion')>
```
Therefore I'm quite sure that the exception case will not raise another error.

Reply:
As I know our daily env. adopts python 3.10 which has different behavior on `parse_version`
```
Python 3.10.14 (main, Mar 24 2024, 08:01:04) [GCC] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from pkg_resources import parse_version
<stdin>:1: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
>>> parse_version("foobar-noversion")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/albin/Workspace/albinsun/tests/.tox/py310/lib/python3.10/site-packages/pkg_resources/_vendor/packaging/version.py", line 200, in __init__
    raise InvalidVersion(f"Invalid version: '{version}'")
pkg_resources._vendor.packaging.version.InvalidVersion: Invalid version: 'foobar-noversion'
```

Reply:
`LegacyVersion` had been dropped in https://github.com/pypa/packaging/pull/407